### PR TITLE
Convert DictionaryExpr to not use tc.callWitness() or generate a SemanticExpr.

### PR DESCRIPTION
**Pull request by pschuh**
For reference, everything else except string interpolation has been
migrated to this new form.

### CODE REVIEW

**Comment by pschuh** [comment]
@swift-ci please test

File: lib/SILGen/SILGenExpr.cpp [hunk 1]
```diff
@@ -3671,15 +3670,20 @@ visitMagicIdentifierLiteralExpr(MagicIdentifierLiteralExpr *E, SGFContext C) {
...
+RValue RValueEmitter::visitCollectionExpr(CollectionExpr *E, SGFContext C) {
   auto loc = SILLocation(E);
   ArgumentScope scope(SGF, loc);
 
   // CSApply builds ArrayExprs without an initializer for the trivial case
   // of emitting varargs.
   CanType arrayType, elementType;
   if (E->getInitializer()) {
-    elementType = E->getElementType()->getCanonicalType();
+    if (auto* arrayExpr = dyn_cast<ArrayExpr>(E)) {
```
Comment (slavapestov) [inline lib/SILGen/SILGenExpr.cpp]: 'auto *'

**Review by slavapestov — APPROVED** [review]
Nice! Is `CollectionExpr::getSemanticExpr()` now dead too?

**Comment by pschuh** [comment]
@swift-ci please test

**Comment by pschuh** [comment]
Yes, I think InterpolatedStringExpr is the last Expr that uses the SemanticExpr solution.

File: lib/Sema/CSDiag.cpp [hunk 2]
```diff
@@ -1426,10 +1413,6 @@ namespace {
     // we go digging through failed constraints, and expect their locators to
     // still be meaningful.
     ~ExprTypeSaverAndEraser() {
-      for (auto CSE : CollectionSemanticExprs)
```
Comment (slavapestov) [inline lib/Sema/CSDiag.cpp]: I love that ExprTypeSaverAndEraser becomes simpler as well (ideally it wouldn't exist; there's a lot of refactoring before we can do that though!)

**Pull request merged**